Ref #29444 -- Add database defaults

codingjoe opened the pull request.

codingjoe commented:
@felixxm I updated the release note to include mention Oracle support for database default. I believe this concludes issue 29444. It's been plenty fun :) Let me know if you have any change requests.

codingjoe commented:
Ref #11568

codingjoe commented:
@apollo13, @charettes  maybe you would be interested in reviewing this too, since you have been very involved with the `returning` API changes and it all kind of lead to this ;)

auvipy commented:
what about migrations?

codingjoe commented:
> what about migrations?

Can you elaborate a bit more what you mean? This is a rather small change, I am happy to put in more effort to move this over the finish line.

charettes commented on django/db/models/fields/__init__.py:
```diff
@@ -738,7 +738,8 @@ def db_returning(self):
         """
-        return False
+        from django.db.models.expressions import BaseExpression
+        return self.has_default() and isinstance(self.default, BaseExpression)
```
We usually use duck typing for such cases

```suggestion
        return self.has_default() and hasattr(self.default, 'as_sql')
```

charettes commented on tests/queries/test_db_returning.py:
```diff
@@ -49,3 +49,12 @@ def test_bulk_insert(self):
+        obj = ReturningModel.objects.create()
+        self.assertIsInstance(obj.created, datetime.datetime)
+        self.assertIsInstance(obj.year, int)
+        self.assertEquals(obj.pi, 3.14159265358979)
```
[`assertEquals` is a deprecated alias](https://docs.python.org/3/library/unittest.html?highlight=assertequals#deprecated-aliases)


```suggestion
        self.assertEqual(obj.pi, 3.14159265358979)
```

codingjoe commented:
> I'd like to see more tests around invalid expression such as ones that span joins, aggregation, window functions. I think we'll want to add checks for that.

Actually, most of them would work. However, since we do not allow `F` expressions in insert, you will get a ValueError:
```
ValueError: Failed to insert expression "Max(Col(queries_returningmodel, queries.ReturningModel.id))" on queries.ReturningModel.max_id. F() expressions can only be used to update, not to insert.
```

Besides it being borderline impossible to implement a comprehensive check, without refactoring all expressions, I don't even know if it's wise to add a one. See strictly speaking any value is allowed and most that don't work should. However, if they don't you will get an error. Analog to that, you can provide a callable, that returns a completely different type than the field at hand. We don't have a check for that, and that might be a good thing.

I don't know I am torn about this. I am happy to add tests tho. But I might need a bit more input before I can progress.

> Also primary key interactions should be tested (e.g. `(primary_key=True , default=Func('uuid'))`) and the documentation should make it clear that this is not a `db_default` option that results in using the SQL `DEFAULT expr` at the table definition level. We already get report of users surprised Django doesn't use `DEFAULT` when only Python callable are supported and I feel this blurs the line even more.

I added a note about the behavior. There already is a test for a primary key, I didn't use UUID, since this the functions are different across Oracle and Postgres and in the latter the `uuid-ossp` extension is required, which I don't want to enable it the suite.

pope1ni commented on docs/ref/models/fields.txt:
```diff
@@ -364,6 +364,25 @@ The default value is used when new model instances are created and a value
+    from django.db import models
+    from django.db.models.functions import ExtractYear, Now, Pi
+
+    class CreatedModel(models.Model)
```
Maybe we should name this `ReturningModel` and keep the documentation and test model aligned?

pope1ni commented on docs/ref/models/fields.txt:
```diff
@@ -364,6 +364,25 @@ The default value is used when new model instances are created and a value
+    The Query Expressions are not stored in the database as default values,
+    but functions are passed as values to the database during inserts and
+    updates. Therefore, the defaults will not be available should you perform
+    any raw SQL inserts or updates and must be provided explicitly too.
```
…any raw SQL ``INSERT`` or ``UPDATE``; you must explicitly provide an equivalent SQL expression. You will not be able to use the ``DEFAULT`` or ``DEFAULT VALUES`` keywords to populate columns for fields using query expressions.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

pope1ni commented on tests/queries/test_db_returning.py:
```diff
@@ -57,6 +57,6 @@ def test_now(self):
         self.assertIsInstance(obj.created, datetime.datetime)
         self.assertIsInstance(obj.year, int)
-        self.assertEqual(obj.pi, 3.14159265358979)
+        self.assertEqual(f"{obj.pi:.2f}", "3.14")  # decimal precision varies
```
Use `self.assertAlmostEqual()` rather than formatting as a string.

pope1ni commented on django/db/backends/base/schema.py:
```diff
@@ -223,7 +223,11 @@ def column_sql(self, model, field, include_default=False):
             if default_value is not None:
-                if self.connection.features.requires_literal_defaults:
+                if hasattr(default_value, 'as_sql'):
+                    s, p = default_value.as_sql(self, self.connection)
```
Use `sql, params =` instead of `s, p =` to match the typical naming used elsewhere in the code.

pope1ni commented on docs/ref/models/fields.txt:
```diff
@@ -364,6 +364,28 @@ The default value is used when new model instances are created and a value
+    The query expressions are not stored in the database as column default
+    values, but expressions are passed directly to the database during inserts
+    and updates. Therefore, the defaults will not be available should you
+    perform any raw SQL `INSERT` or `UPDATE`; you must explicitly provide an
```
Please use double backticks around `INSERT`, `DEFAULT`, etc.

pope1ni commented on docs/ref/models/fields.txt:
```diff
@@ -364,6 +364,28 @@ The default value is used when new model instances are created and a value
+    from django.db import models
+    from django.db.models.functions import ExtractYear, Now, Pi
+
+    class Returning(models.Model)
```
`Returning` → `ReturningModel`

codingjoe commented:
buildbot, test on oracle.

auvipy commented on docs/ref/models/fields.txt:
```diff
@@ -364,6 +364,28 @@ The default value is used when new model instances are created and a value
 isn't provided for the field. When the field is a primary key, the default is
 also used when the field is set to ``None``.
 
+.. versionadded:: 3.1
```
3.2

codingjoe commented on docs/ref/models/fields.txt:
```diff
@@ -364,6 +364,28 @@ The default value is used when new model instances are created and a value
 
+.. versionadded:: 3.1
+
+The Oracle and PostgreSQL database backends support
```
@felixxm did someone add support for the `db_retuning` feature that I wrote?

felixxm replied:
Yes, see ticket-31403 and 93ed71e05802a47774b52503cdc3442686d686c1.

codingjoe replied:
Cool, thanks!

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
@felixxm excellent points. I added plenty more tests especially around migrations. I also added another field check to warn users about column references. Let me know if you have any more feedback. Best, Joe

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
@auvipy you seem very excited, would you care to review the patch to move things along a little quicker? This is a rather complex patch, so every pair eye helps.

InvalidInterrupt commented on django/db/backends/oracle/schema.py:
```diff
@@ -149,6 +149,9 @@ def _generate_temp_name(self, for_name):
     def prepare_default(self, value):
+        if hasattr(value, 'as_sql'):
+            sql, params = self.prepare_param(value)
+            return sql % params
```
Should some effort be made to `quote_value()` the params before construction?

codingjoe replied:
No, since this is a SQL expression like `NULL` or `CURRENT_TIMESTAMP` not a value.

InvalidInterrupt replied:
If a user did something unusual like `CharField(default=Value(";"))` would that not pass through this function and become a literal semicolon (which would probably not have been what the user expected)?

charettes replied:
@InvalidInterrupt is right, this should map `params` using `quote_value` just like it's done on other backends.

codingjoe replied:
After careful consideration, I must continue to object. See, this method is Oracle-specific, it isn't implemented nor called for other backends but Oracle. With the main reason that Oracle requires literal defaults. For that matter, quoting all values will simply not work, since it means quoting db expressions, which breaks the whole feature.

However, I do see the concern about `Values` expressions. I will add an exception for that. I hope that is satisfactory. If not I am open to solutions.

codingjoe commented on django/db/backends/base/schema.py:
```diff
@@ -41,7 +42,7 @@ def _related_non_m2m_objects(old_field, new_field):
 
 
-class BaseDatabaseSchemaEditor:
+class BaseDatabaseSchemaEditor(SQLCompiler):
```
Hm… yes, you are correct, I only needed the `prepare_param` method. However, I also believe the `SchemaEditor` should be subclass of `SQLCompiler` just for consistency. After all, that's what it does, right? However, this certainly is up for debate and might be out of scope and will slow down the release of this patch. I will go with your suggestions.

codingjoe commented on django/db/backends/base/schema.py:
```diff
@@ -118,6 +119,17 @@ def __exit__(self, exc_type, exc_value, traceback):
+        vendor_impl = getattr(node, 'as_' + self.connection.vendor, None)
+        if vendor_impl:
+            sql, params = vendor_impl(self, self.connection)
+        else:
+            try:
+                sql, params = node.as_sql(self, self.connection)
+            except AttributeError:
+                sql, params = None, (node,)
```
Yes it does, I was going for EAFP, but your suggestions makes more sense here.

codingjoe commented on docs/ref/models/fields.txt:
```diff
@@ -363,6 +363,28 @@ The default value is used when new model instances are created and a value
 
+.. versionadded:: 3.2
+
+The MariaDB, Oracle and PostgreSQL database backends support
```
Yes, good catch, it's 10.5+

codingjoe commented on django/db/models/base.py:
```diff
@@ -899,7 +899,7 @@ def _save_table(self, raw=False, cls=None, force_insert=False,
             if results:
                 for value, field in zip(results[0], returning_fields):
-                    setattr(self, field.attname, value)
+                    setattr(self, field.attname, field.to_python(value))
```
🤫 yes, it's a bugfix, I don't want to rebase have yet another patch pending somewhere 😭

codingjoe commented on django/db/backends/base/schema.py:
```diff
@@ -129,17 +141,25 @@ def execute(self, sql, params=()):
+        prepared_params = None
+        if params is not None:
+            prepared_params = [self.prepare_param(p) for p in params]
+            params_sql = tuple(filter(None, (p[0] for p in prepared_params)))
+            if params_sql:
+                sql %= params_sql
```
Since fields are only added one field a time, we are safe here.

codingjoe commented:
@InvalidInterrupt thanks for the thorough review, it really helped to get this forward. If you are happy with my changes let me know and I will squash my changes to get this ready for check in.

codingjoe commented:
buildbot, test on oracle.

charettes commented on django/db/backends/base/schema.py:
```diff
@@ -118,6 +118,17 @@ def __exit__(self, exc_type, exc_value, traceback):
     # Core utility functions
 
+    def prepare_param(self, node):
+        compiler = self.connection.ops.compiler("SQLCompiler")(query=None, connection=self.connection, using=None)
+        vendor_impl = getattr(node, 'as_' + self.connection.vendor, None)
+        if vendor_impl:
+            sql, params = vendor_impl(compiler, self.connection)
+        elif hasattr(node, 'as_sql'):
+            sql, params = node.as_sql(compiler, self.connection)
+        else:
+            sql, params = None, (node,)
+        return sql, tuple(params)
```
We expect compileable to always have an `as_sql` method. e.g. see `effective_default`'s implementation.

```suggestion
        if not hasattr(node, 'as_sql'):
            return None, (node,)
        compiler = self.connection.ops.compiler('SQLCompiler')(
            query=None, connection=self.connection, using=None
        )
        return compiler.compile(node)
```

charettes commented on django/db/backends/base/schema.py:
```diff
@@ -129,17 +140,25 @@ def execute(self, sql, params=()):
             )
         # Account for non-string statement objects.
         sql = str(sql)
+        prepared_params = None
+        if params is not None:
+            prepared_params = [self.prepare_param(p) for p in params]
+            params_sql = tuple(filter(None, (p[0] for p in prepared_params)))
+            if params_sql:
+                sql %= params_sql
+            prepared_params = tuple(q for p in prepared_params for q in p[1])
+
```
Feels like this should not be handled by `execute` but by its callers that need to pass unprepared params.

Was this to avoid updating `create_model`, `add_field`, and `alter_field` paths?

codingjoe replied:
Should it, I don't know? It would cause a lot of duplication? You are Mr. Database, so I'd follow your advice. Any concrete suggestions where to put this?

codingjoe commented:
buildbot, test on oracle.

Ian-Foote commented on django/db/backends/base/schema.py:
```diff
@@ -121,6 +121,15 @@ def __exit__(self, exc_type, exc_value, traceback):
     # Core utility functions
 
+    def prepare_param(self, node):
+        sql, params = None, (node,)
+        if hasattr(node, 'as_sql'):
+            compiler = self.connection.ops.compiler('SQLCompiler')(
+                query=None, connection=self.connection, using=None
+            )
+            sql, params = compiler.compile(node)
+        return sql, tuple(params)
```
I think it's a bit neater to invert the check for `node.as_sql` and return early:

```suggestion
        if not hasattr(node, 'as_sql'):
            return None, (node,)
        compiler = self.connection.ops.compiler('SQLCompiler')(
            query=None, connection=self.connection, using=None
        )
        sql, params = compiler.compile(node)
        return sql, tuple(params)
```

codingjoe replied:
hm… this will increase mccabe complexity and I also personally don't think a negative if statement will increase readability. If you don't mind, I will keep it as i.s

cjerdonek replied:
I agree with @Ian-Foote. For example, the first line isn't used in the case that `hasattr(node, 'as_sql')` is true.

Ian-Foote commented on tests/schema/tests.py:
```diff
@@ -2832,8 +2835,7 @@ def test_func_index_unsupported(self):
         index = Index(Lower('nonexistent'), name='func_nonexistent_idx')
         msg = (
-            "Cannot resolve keyword 'nonexistent' into field. Choices are: "
-            "height, id, name, uuid, weight"
```
Why has this line been removed?

codingjoe replied:
I just added a field, but I'll fix the message, good point.

InvalidInterrupt commented on django/db/backends/oracle/schema.py:
```diff
@@ -152,6 +153,11 @@ def _generate_temp_name(self, for_name):
     def prepare_default(self, value):
+        if hasattr(value, 'as_sql'):
+            sql, params = self.prepare_param(value)
+            if isinstance(value, Value):
+                params = tuple(self.quote_value(p) for p in params)
+            return sql % params
```
I believe `params` should no longer contain any more `Expression` objects, and I ran the test suite with this change and didn't see any related failures. Trying to special case `Value` (and other expressions like it) seems like a losing battle because they may be nested inside other expressions.
```suggestion
            return sql % tuple(self.quote_value(p) for p in params)
```

codingjoe replied:
Yes, that seems to work. Thanks

codingjoe commented:
buildbot, test on oracle.

luzfcb commented:
> I added a note about the behavior. There already is a test for a primary key, I didn't use UUID, since this the functions are different across Oracle and Postgres, and in the latter, the `uuid-ossp` extension is required, which I don't want to enable it the suite.

PostgreSQL 13 now includes the `gen_random_uuid()` function without requires the `uuid-ossp` extension and in teory the `RandomUUID` can be used without run `CryptoExtension` on the migration . https://www.postgresql.org/docs/13/functions-uuid.html

That said, A small brainstorm, ignore if it doesn't make sense

I'm especially curious to know what django's behavior will be on INSERT (create/bulk_create) and/or UPDATE, when we have a field with `unique=True` and the `default` is an expression that runs some database function (like `gen_random_uuid()`) and for some reason generated an item with a value that already exists in the column, that is, generating a unique constraint error.



1 - Will Django just raise the exception?
2 - Will Django try a few times again before raising an exception?  (Maybe an Expressions API has some way of indicating that the expression supports this hypothetical feature)

In short, Django would help somehow to be able to recover from this type of error on its own, and if not, how should an application writer handle handling these errors?


The most common use case I think would be use UUIDField as primary key or a value to using externally (like rest apis):

```python
from django.db import models
from django.contrib.postgres.functions import RandomUUID

class Question(models.Model):
    external_id = models.UUIDField(unique=True, null=False, default=RandomUUID, db_index=True, editable=False)
```

codingjoe commented:
@luzfcb thank you for your concern, but I believe this isn't the best place to discuss this. I would suggest opening a new ticket for this. But as a quick info, thus far the ORM doesn't handle PG extension dependencies. If you use fields or functions that require extensions, you will need to install those manually.

cjerdonek commented on django/db/backends/base/schema.py:
```diff
@@ -121,6 +121,15 @@ def __exit__(self, exc_type, exc_value, traceback):
     # Core utility functions
 
+    def prepare_param(self, node):
+        sql, params = None, (node,)
```
This assignment and tuple creation /  unpacking, etc. isn't used in the case that `hasattr` is true. I also noted this below, but the conversation is marked resolved.

cjerdonek commented on django/db/backends/base/schema.py:
```diff
@@ -132,17 +141,25 @@ def execute(self, sql, params=()):
         # Account for non-string statement objects.
         sql = str(sql)
+        prepared_params = None
+        if params is not None:
```
Can this be `if params`? Does the `if` block need to execute if `params` is an empty tuple (the method default)?

cjerdonek reviewed (commented):
A couple comments.

felixxm commented:
@codingjoe Thanks for all your efforts :1st_place_medal: and sorry for the late reply :broken_heart: . Unfortunately, we need to make few things to make it reviewable again:
- rebase from the `main` branch,
- apply `black`, and
- check failing tests.

I hope it makes sense.

codingjoe commented:
@felixxm  😉 

![image](https://user-images.githubusercontent.com/1772890/157908823-86783c47-5524-4253-9241-e458b77200f5.png)

felixxm commented:
> @felixxm I know the drill wink

I can promise a detailed review before Django 4.1 feature freeze, if it will be reviewable again.

nessita commented:
@codingjoe I have rebased and fixed conflicts and code style. There were a couple of changes I wasn't sure about so you'd need to review and fix accordingly:

1. Implementation of `db_returning` changed since you made this branch, so the part `and hasattr(self.default, "as_sql")` needs proper placing.
2. Many tests fail using sqlite because of the `raise NotImplementedError` in `NativeUUID4.as_sql`, so this needs fixing.

Hopefully you can pick up this PR from here! Let us know. Thanks!

nessita commented:
I also had to revert this change because lots of tests (specifically in `custom_pk` where failing:

```diff
diff --git a/django/db/models/base.py b/django/db/models/base.py
index 0f8af9f9204e3..11b7f04dd34ef 100644
--- a/django/db/models/base.py
+++ b/django/db/models/base.py
@@ -872,7 +872,7 @@ def _save_table(self, raw=False, cls=None, force_insert=False,
             results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
             if results:
                 for value, field in zip(results[0], returning_fields):
-                    setattr(self, field.attname, value)
+                    setattr(self, field.attname, field.to_python(value))
         return updated
 
     def _do_update(self, base_qs, using, pk_val, values, update_fields, forced_update):
```

felixxm commented:
Closing due to inactivity.

codingjoe commented:
I'll try to get to it this Xmas 🎄
